### PR TITLE
Use non-broadcasting positions in `multiReductionInliner`

### LINKED ISSUE #2241: transformRFactor::runReplay asserts on axis out of range.
### 🐛 Describe the bug

Running into codegen issue:
`RuntimeError: Rfactor replay received an axis outside the number of dims in the tensor, acceptable inclusive range is -4 to 3`

fusion IR from the failing tests
```
inputs: 
T0_g[ bS0{1} ] int64_t
T1_g[ iS1{i2}, iS2{i3}, bS3{1}, iS4{i4} ] double
T2_g[ bS5{1}, bS6{1}, bS7{1}, bS8{1} ] int64_t
outputs: 
T13_g[ bS41{1} ] double


T3_l[ rS9{i2}, rS10{i3}, bS11{1}, rS12{i4} ]
   = reduction( T1_g[ iS1{i2}, iS2{i3}, bS3{1}, iS4{i4} ], op = add, initial value = double(0), allreduce = false )
T4_l[ bS13{1}, bS14{1}, bS15{1}, bS16{1} ]
   = broadcast( T3_l[ rS9{i2}, rS10{i3}, bS11{1}, rS12{i4} ] )
T5_l[ bS17{1}, bS18{1}, bS19{1}, bS20{1} ]
   = (double)(T2_g[ bS5{1}, bS6{1}, bS7{1}, bS8{1} ]);
T6_l[ bS21{1}, bS22{1}, bS23{1}, bS24{1} ]
   = T4_l[ bS13{1}, bS14{1}, bS15{1}, bS16{1} ]
   / T5_l[ bS17{1}, bS18{1}, bS19{1}, bS20{1} ];
T7_l[ iS25{i2}, iS26{i3}, bS27{1}, iS28{i4} ]
   = T1_g[ iS1{i2}, iS2{i3}, bS3{1}, iS4{i4} ]
   - T6_l[ bS21{1}, bS22{1}, bS23{1}, bS24{1} ];
T8_l[ iS29{i2}, iS30{i3}, bS31{1}, iS32{i4} ]
   = T7_l[ iS25{i2}, iS26{i3}, bS27{1}, iS28{i4} ]
   * T7_l[ iS25{i2}, iS26{i3}, bS27{1}, iS28{i4} ];
T9_l[ iS33{i2}, iS34{i3}, iS35{i4} ]
   = squeeze( T8_l[ iS29{i2}, iS30{i3}, bS31{1}, iS32{i4} ] )
T10_l[ rS36{i2}, rS37{i3}, rS38{i4} ]
   = reduction( T9_l[ iS33{i2}, iS34{i3}, iS35{i4} ], op = add, initial value = double(0), allreduce = false )
T12_l[ bS40{1} ]
   = broadcast( T10_l[ rS36{i2}, rS37{i3}, rS38{i4} ] )
T11_l[ bS39{1} ]
   = (double)(T0_g[ bS0{1} ]);
T13_g[ bS41{1} ]
   = T12_l[ bS40{1} ]
   / T11_l[ bS39{1} ];
}

} //Segmented_Fusion
```

python repro:
```
import torch
  
def fn(t6, t15, t20):
    sample_total = torch.sum(t15, [0, 1, 2, 3], keepdim=True)
    sample_mean = torch.div(sample_total, t20)
    x = torch.sub(t15, sample_mean)
    input = torch.mul(x, x)
    total = torch.sum(input, [0, 1, 2, 3], keepdim=False)
    t7 = torch.div(total, t6)
    return t7

t6 = torch.tensor([15], dtype=torch.long).cuda()
t15 = torch.rand([3, 2, 1, 2], dtype=torch.float32).cuda().double()
t20 = torch.tensor([12], dtype=torch.long).cuda().expand([1, 1, 1, 1])

jit_fn = torch.jit.script(fn)
for i in range(5):
    print(i)
    jit_fn(t6, t15, t20)
```

### Versions

Verified on devel:
```
commit acb42b987108d6965959a2b99aba76b597152ffb (HEAD -> devel, origin/devel, origin/HEAD)
Author: Naoya Maruyama <naoyam@users.noreply.github.com>
Date:   Thu Dec 1 08:42:48 2022 -0800

    Fix #2230 (#2231)
```

This is impacting upstream bump PR: https://github.com/pytorch/pytorch/pull/89460

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes #2241.

Because in `checkPatternEquivalence`, broadcast is ignored, we might end up having multiple reductions with pattern equivalence but have different number of broadcasts, so the position in the reference tensor is not necessary the same as the position in other reduction TVs. To handle this,  this PR takes advantage of the fact that broadcast dim can not be reduced, and changes `multiReductionInliner` to use positions in non-broadcast axes.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 1]
```diff
@@ -172,10 +172,13 @@ Val* replaceValInIndexVal(
     Val* index,
     const std::unordered_map<Val*, Val*>& replacement_map);
 
-// Makes rfactor generic with reduction ops and Welford
+// Makes rfactor generic with reduction ops and Welford.
+// Note that the axes must be positions excluding broadcast. For example,
+// if you have a tensor T1[b, rS1, rS2, rS3], and you want to rFactor axis rS2
+// and rS3, you will need to use rfactorHelper(T1, {1, 2}) instead
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: Is this change really necessary? I think it makes most sense to have the same parameter semantics as `reductionOp`.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: `rfactorHelper` converts the position parameters with no broadcast to the raw positions, so can it be done by the scheduler instead?

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: I agree, will change

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Changed. It is much cleaner now. Thank you for your suggestion.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for fixing the bug.

**Pull request merged**